functions/src/sections/Section.service.ts

```diff
@@ -26,24 +26,57 @@ export class SectionsService {
     code: string
   ): Promise<Seat[]> {
     // get term, subject, code to crn mappings from db.
-    const doc = await db.sectionMappings
+    const mapping =
+      (await this.getSectionMapping(term, subject, code)) ||
+      (await this.setSectionMapping(term, subject, code));
+
+    if (mapping) {
+      return await Promise.all(
+        mapping.crns.map(async (crn) => {
+          const seat = await UVicCourseScraper.getSectionSeats(term, crn);
+          return { ...seat, crn, date: new Date(Date.now()) };
+        })
+      );
+    }
+    throw new SectionNotFoundError('Seats Not Found');
+  }
+
+  public async getSectionMapping(
+    term: string,
+    subject: string,
+    code: string
+  ): Promise<CourseMapping | undefined> {
+    const doc = await db.courseMappings
       .doc(SectionsService.constructSectionKey(term, subject, code))
       .get();
 
-    const data = doc.data();
-
-    if (!data) {
-      throw new SectionNotFoundError('Seats Not Found');
+    const t = doc.data()?.retrieveAt?.getTime();
+    // if retrivedAt exists and it was retieved within 30 minutes
+    if (t && t + 1000 * 1800 > Date.now()) {
+      return undefined;
     }
+    return doc.data();
+  }
 
-    const seats = await Promise.all(
-      data.crns.map(async (crn) => {
-        const seat = await UVicCourseScraper.getSectionSeats(term, crn);
-        return { ...seat, crn };
-      })
-    );
-
-    return seats;
+  public async setSectionMapping(
+    term: string,
+    subject: string,
+    code: string
+  ): Promise<CourseMapping | void> {
+    try {
+      const sections = await UVicCourseScraper.getCourseSections(
+        term,
+        subject,
+        code
+      );
+      if (sections.length > 0) {
+        const crns = sections.map(({ crn }) => crn);
+        await db.courseMappings
+          .doc(SectionsService.constructSectionKey(term, subject, code))
+          .set({ crns });
```

Comment:
shouldn't retrievedAt be written to database?

Reply:
should we have any error handling in case this throws? at least log the error? I'm open for discussion

Reply:
Fixed the first comment, nice catch and thanks. Second, I did add some logging as it's better than nothing now. I haven't had time to looking into our monitoring ability of app yet but that's a whole other thing and out of scope for this task.

Reply:
fair point

Reply:
I think it's a matter of adding middleware in Express but like I said, haven't found a good way.